File: Blind75/133. Clone Graph/133. Clone Graph.cpp
class Solution {
public:
    unordered_map<Node*, Node*> clone;
    Node* cloneGraph(Node* node) {
        //DFS
        if(!node) return NULL;
        if(!clone.count(node)){
            clone[node] = new Node(node->val);
            for(auto it:node->neighbors){
                clone[node]->neighbors.push_back(cloneGraph(it));
            }
        }
        return clone[node];
        /* //BFS
        if(!node) return NULL;
        queue<Node*> q;
        clone[node] = new Node(node->val);
        q.push(node);

        while(!q.empty()){
            Node* curr = q.front();
            q.pop();
            for(auto it:curr->neighbors){
                if(!clone.count(it)){
                    clone[it] = new Node(it->val);
                    q.push(it);
                }
                clone[curr]->neighbors.push_back(clone[it]);
            }
        }

        return clone[node]; */
    }

File: Blind75/133. Clone Graph/133. Clone Graph.cs
public class Solution {
    Dictionary<Node,Node> clone = new Dictionary<Node,Node>();
    public Node CloneGraph(Node node) {
        //BFS
        if(node==null) return null;
        Queue<Node> q = new Queue<Node>();
        clone[node] = new Node(node.val);
        q.Enqueue(node);

        while(q.Count>0){
            Node curr = q.Peek();
            q.Dequeue();
            
            foreach(var it in curr.neighbors){
                if(!clone.ContainsKey(it)){
                    clone[it] = new Node(it.val);
                    q.Enqueue(it);
                }
                clone[curr].neighbors.Add(clone[it]);
            }

        }
        return clone[node];
    }
}
